cvmfs/receiver/catalog_merge_tool.cc

```diff
@@ -0,0 +1,234 @@
+      entry1_(new catalog::DirectoryEntry(*(other.entry1_))),
+      entry2_(other.entry2_ ? new catalog::DirectoryEntry(*(other.entry2_))
+                            : NULL) {}
+
+CatalogMergeTool::ChangeItem::~ChangeItem() {
+  if (entry1_) {
+    delete entry1_;
+    entry1_ = NULL;
+  }
+  if (entry2_) {
+    delete entry2_;
+    entry2_ = NULL;
+  }
+}
+
+CatalogMergeTool::CatalogMergeTool(const std::string& repo_path,
+                                   const std::string& old_root_hash,
+                                   const std::string& new_root_hash,
+                                   const std::string& temp_dir_prefix,
+                                   download::DownloadManager* download_manager,
+                                   manifest::Manifest* manifest)
+    : CatalogDiffTool(repo_path, old_root_hash, new_root_hash, temp_dir_prefix,
+                      download_manager),
+      repo_path_(repo_path),
+      temp_dir_prefix_(temp_dir_prefix),
+      download_manager_(download_manager),
+      manifest_(manifest) {}
+
+CatalogMergeTool::~CatalogMergeTool() {}
+
+bool CatalogMergeTool::Run(const Params& params,
+                           std::string* new_manifest_path) {
+  bool ret = CatalogDiffTool::Run();
+
+  upload::SpoolerDefinition definition(
+      params.spooler_configuration, params.hash_alg, params.compression_alg,
+      params.use_file_chunking, params.min_chunk_size, params.avg_chunk_size,
+      params.max_chunk_size, "dummy_token", "dummy_key");
+  UniquePtr<upload::Spooler> spooler(upload::Spooler::Construct(definition));
+  perf::Statistics stats;
+  const std::string temp_dir = CreateTempDir(temp_dir_prefix_);
+  output_catalog_mgr_ = new catalog::WritableCatalogManager(
+      manifest_->catalog_hash(), repo_path_, temp_dir, spooler,
+      download_manager_, params.entry_warn_thresh, &stats,
+      params.use_autocatalogs, params.max_weight, params.min_weight);
+  output_catalog_mgr_->Init();
+
+  ret &= InsertChangesIntoOutputCatalog();
+
+  ret &= CreateNewManifest(new_manifest_path);
+
+  return ret;
+}
+
+void CatalogMergeTool::ReportAddition(const PathString& path,
+                                      const catalog::DirectoryEntry& entry,
+                                      const XattrList& xattrs) {
+  changes_.push_back(
+      ChangeItem(ChangeItem::kAddition, MakeRelative(path), entry, xattrs));
+}
+
+void CatalogMergeTool::ReportRemoval(const PathString& path,
+                                     const catalog::DirectoryEntry& entry) {
+  changes_.push_back(
+      ChangeItem(ChangeItem::kRemoval, MakeRelative(path), entry));
+}
+
+void CatalogMergeTool::ReportModification(
+    const PathString& path, const catalog::DirectoryEntry& entry1,
+    const catalog::DirectoryEntry& entry2) {
+  changes_.push_back(ChangeItem(ChangeItem::kModification, MakeRelative(path),
+                                entry1, entry2));
+}
+
+bool CatalogMergeTool::InsertChangesIntoOutputCatalog() {
```

Comment:
I haven't even thought about that. I don't know how `.cvmfscatalog` works. This case should be covered in a unit or integration test.